cmd/restic/cmd_backup.go

```diff
@@ -522,3 +544,87 @@ func readExcludePatternsFromFiles(excludeFiles []string) []string {
 	}
 	return excludes
 }
+
+// FilenameCheck is a function that takes a filename and returns a boolean depending on arbitraty check
+type FilenameCheck func(filename string) bool
+
+// genExcludeByFile returns a FilenameCheck which itself returns whether a path
+// should be excluded. The FilenameCheck considers a file to be excluded when
+// it resides in a directory with an exclusion file, that is either specified
+// by excludeIfPresent in the form "filename[:content]".  The returned error is
+// non-nil if the filename component of excludeIfPresent is empty.
+func genExcludeByFile(excludeIfPresent string) (FilenameCheck, error) {
+	fn := func(string) bool { return false }
+	tf, tc := "", ""
+	if excludeIfPresent != "" {
+		colon := strings.Index(excludeIfPresent, ":")
+		switch {
+		default:
+			tf = excludeIfPresent[:colon]
+			tc = excludeIfPresent[colon+1:]
+		case colon < 0:
+			tf = excludeIfPresent
+		case colon == 0:
+			return nil, fmt.Errorf("no name for exclusion tagfile provided")
+		}
+	}
+	if tf != "" {
+		debug.Log("using %q as exclusion tagfile", tf)
+		fn = func(filename string) bool {
+			return isExcludedByFile(filename, tf, tc)
+		}
+	}
+	return fn, nil
+}
+
+// isExcludedByFile interprets filename as a path and returns true if that file
+// is in a excluded directory. A directory is identified as excluded if it contains a
+// tagfile which bears the name specified in tagFilename and starts with header.
+func isExcludedByFile(filename, tagFilename, header string) bool {
+	if tagFilename == "" {
+		return false
+	}
+	dir, base := filepath.Split(filename)
+	if base == tagFilename {
+		return false // do not exclude the tagfile itself
+	}
+	tf := filepath.Join(dir, tagFilename)
+	_, err := os.Stat(tf)
+	if os.IsNotExist(err) {
+		return false
+	}
+	if err != nil {
+		Warnf("could not access exclusion tagfile: %v", err)
+		return false
+	}
+	// when no signature is given, the mere presence of tf is enough reason
+	// to exclude filename
+	if len(header) == 0 {
+		return true
+	}
+	// From this stage, errors mean tagFilename exists but it is malformed.
+	// Warnings will be generated so that the user is informed that the
+	// indented ignore-action is not performed.
+	f, err := os.Open(tf)
+	if err != nil {
+		Warnf("could not open exclusion tagfile: %v", err)
+		return false
+	}
+	defer f.Close()
+	var buf bytes.Buffer
+	_, err = io.CopyN(&buf, f, int64(len(header)))
```

Comment:
This code can be simplified: Use `buf := make([]byte, len(header))` and use `io.ReadFull(f, buf)`.

Reply:
Done.